### PR TITLE
Add CuPy JIT Kernel definition

### PR DESCRIPTION
**Pull request by asi1024**
Part of #4290.

This PR supports a raw kernel interface like as following.
```py
from cupyx import jit

@jit.rawkernel()
def f(x, y, z, n):
    tid = jit.threadIdx.x + jit.blockIdx.x * jit.blockDim.x
    ntid = jit.blockDim.x * jit.gridDim.x
    for i in range(tid, n, ntid):
        z[i] = x[i] + y[i]

n = numpy.uint32(1024)
x = cupy.arange(n)
y = cupy.arange(n)
z = cupy.empty((n,), dtype='l')
f[16, 16](x, y, z, n)
```

### CODE REVIEW

File: cupyx/jit/_compile.py [hunk 1]
```diff
@@ -519,12 +547,21 @@ def _transpile_expr(expr, env):
...
+        index = _transpile_expr(expr.slice, env)
+        if is_constants([value, index]):
+            return Constant(value[index])
+        value = _to_cuda_object(value, env)
+        index = _to_cuda_object(index, env)
+
+        if not isinstance(value.ctype, _types.Array):
+            raise ValueError(f'{value.code} must be Array type.')
+        if value.ctype.ndim != 1:
+            raise NotImplementedError('Not implemented for ndim > 1.')
```
Comment (asi1024) [inline cupyx/jit/_compile.py:560]: Currently, only 1-dim arrays are supported, but we are planning to support also multi-dimension arrays in the next PR.

File: tests/cupyx_tests/jit_tests/test_raw.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        @jit.rawkernel()
+        def f(x, y, size):
+            tid = jit.threadIdx.x + jit.blockDim.x * jit.blockIdx.x
+            ntid = jit.blockDim.x * jit.gridDim.x
+            for i in range(tid, size, ntid):
+                y[i] = x[i]
+
+        x = testing.shaped_random((1024,), dtype=numpy.int32, seed=0)
+        y = testing.shaped_random((1024,), dtype=numpy.int32, seed=1)
+        f[5, 6](x, y, numpy.uint32(1024))
```
Comment (asi1024) [inline tests/cupyx_tests/jit_tests/test_raw.py:55]: `tid` and `ntid` is of `uint32` datatype and `range` requires arguments of the same datatype. So, we need to cast `size` to uint32 here. Is it inconvenient?

**Comment by leofang** [comment]
Question: is there a plan to support dynamic allocated shared memory & explicit stream argument? Like
```python
f[16, 16, 256, s](x, y, z, n)  # allocating 256 bytes of shared memory, run the kernel on stream s
```
This is more or less in line with #3684.

File: tests/cupyx_tests/jit_tests/test_raw.py (hunk 2 above)
Comment (emcastillo) [reply tests/cupyx_tests/jit_tests/test_raw.py:55]: I think its reasonable.

File: tests/cupyx_tests/jit_tests/test_raw.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+import unittest
+import numpy
+
+import cupy
+from cupyx import jit
+from cupy import testing
+
+
+class TestRaw(unittest.TestCase):
```
Comment (emcastillo) [inline tests/cupyx_tests/jit_tests/test_raw.py:9]: let's remove unittest as discussed yesterday and it is good to go :)

**Comment by emcastillo** [comment]
Jenkins, test this please

**Review by emcastillo — APPROVED** [review]

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by jakirkham** [comment]
Very cool! Thanks for working on this. Looking forward to trying CuPy JIT 😄

**Comment by asi1024** [comment]
@leofang Yes, we are planning to support dynamic allocated shared memory and explicit stream argument, but their interface is still under discussion.

**Pull request merged**